### PR TITLE
Remove method `qasm` from `Instruction` and all of its subclasses

### PR DESCRIPTION
**Pull request by jlapeyre**
This PR deprecates all methods named `qasm` for `Instruction` and subclasses. Instead the OpenQASM 2 strings are produced in `QuantumCircuit.qasm`.

~~This does little more than remove the method `Instruction.qasm` and fix up all the places that call it to preserve existing behavior.~~

~~Note that this method has been removed, rather than deprecated. In fact this new code relies on `hasattr` not finding this method. When deprecating, this would have to be fixed up.~~

EDIT: The items below have been added to this PR

Looking at
* #10361

and related discussion, it seems that what we really want is to push this PR further by removing all methods called `qasm` from subclasses of `Instruction`.

### CODE REVIEW

File: qiskit/circuit/library/standard_gates/x.py [hunk 1]
```diff
@@ -578,7 +587,10 @@ def qasm(self):
         old_name = self.name
         self.name = "c3sqrtx"
         try:
-            return super().qasm()
+            # pylint: disable=cyclic-import
+            from qiskit.circuit.quantumcircuit import _instruction_qasm2
+
+            return _instruction_qasm2(self)
         finally:
             self.name = old_name
```
Comment (jakelishman) [inline qiskit/circuit/library/standard_gates/x.py:595]: I think it's fine for this to continue to call up to `super().qasm()` - we're already in a deprecated context.

File: qiskit/circuit/quantumcircuit.py [hunk 2]
```diff
@@ -1737,16 +1740,24 @@ def qasm(
             elif operation.name == "barrier":
                 qargs = ",".join(bit_labels[q] for q in instruction.qubits)
                 instruction_qasm = "barrier;" if not qargs else f"barrier {qargs};"
+            elif hasattr(operation, "has_qasm2") and not operation.has_qasm2():
+                raise CircuitError(
+                    f"No OpenQASM 2 representation is defined for operation {operation.name}"
+                )
```
Comment (jakelishman) [inline qiskit/circuit/quantumcircuit.py:1746]: `operation.has_qasm2`? We shouldn't be adding new magic interfaces; we're trying to get rid of them.

File: qiskit/circuit/quantumcircuit.py [hunk 3]
```diff
@@ -2817,7 +2828,7 @@ def _unroll_param_dict(
         self, value_dict: Mapping[Parameter, ParameterValueType]
     ) -> dict[Parameter, ParameterValueType]:
         unrolled_value_dict: dict[Parameter, ParameterValueType] = {}
-        for (param, value) in value_dict.items():
+        for param, value in value_dict.items():
```
Comment (jakelishman) [inline qiskit/circuit/quantumcircuit.py:2831]: Can we revert unrelated formatting changes?

File: qiskit/circuit/quantumcircuit.py [hunk 4]
```diff
@@ -1737,16 +1740,24 @@ def qasm(
...
+                if operation.name == "c3sx":
+                    operation.name = "c3sqrtx"
+                name_param = _instruction_qasm2(operation)
+                if operation.name == "c3sqrtx":
+                    operation.name = "c3sx"
                 bits_qasm = ",".join(
                     bit_labels[j] for j in itertools.chain(instruction.qubits, instruction.clbits)
                 )
-                instruction_qasm = f"{operation.qasm()} {bits_qasm};"
+                instruction_qasm = f"{name_param} {bits_qasm};"
```
Comment (jakelishman) [inline qiskit/circuit/quantumcircuit.py:1760]: The ordering of these calls doesn't look right - `_qasm2_define_custom_operation` is what actually defines the operation and it's true recursive.  The special handling should be in there if anything - it already has logic to transform one operation into another for use as the definition and call point.

File: qiskit/circuit/quantumcircuit.py [hunk 5]
```diff
@@ -5199,3 +5211,22 @@ def _bit_argument_conversion_scalar(specifier, bit_sequence, bit_set, type_):
...
+        )
+    if operation.condition is not None:
+        if not isinstance(operation.condition[0], ClassicalRegister):
+            raise QasmError(
+                "OpenQASM 2 can only condition on registers, but got '{operation.condition[0]}'"
+            )
+        name_param = (
+            "if(%s==%d) " % (operation.condition[0].name, operation.condition[1]) + name_param
+        )
+    return name_param
```
Comment (jakelishman) [inline qiskit/circuit/quantumcircuit.py:5232]: I get that it was pre-existing, but while we're at it, can we call `name_param` here (and all the places in `QuantumCircuit.qasm`) something that actually describes it - it's not just a name, it's a potentially conditional call with parameters.

File: qiskit/extensions/hamiltonian_gate.py [hunk 6]
```diff
@@ -116,6 +117,17 @@ def _define(self):
         qc._append(UnitaryGate(self.to_matrix()), q[:], [])
         self.definition = qc
 
+    def has_qasm2(self):
+        """Return `False` because no QASM2 representation is defined for HamiltonianGate."""
+        return False
+
```
Comment (jakelishman) [inline qiskit/extensions/hamiltonian_gate.py:123]: We shouldn't add more magic interfaces, and in fact this gate shouldn't be doing anything like this.  If you just remove this handling, I think the exporter will actually succeed in outputting the gate.

File: test/python/circuit/test_circuit_qasm.py [hunk 7]
```diff
@@ -266,6 +266,11 @@ def test_c3sxgate_roundtrips(self):
         parsed = QuantumCircuit.from_qasm_str(qasm)
         self.assertIsInstance(parsed.data[0].operation, C3SXGate)
 
+    def test_c3sxgate_qasm_deprecation_warning(self):
+        """Test deprecation warning for C3SXGate."""
+        with self.assertWarnsRegex(DeprecationWarning, r"Correct exporting to OpenQASM 2"):
+            C3SXGate().qasm()
+
```
Comment (jakelishman) [inline test/python/circuit/test_circuit_qasm.py:273]: I'm not sure why we need this, but I guess it doesn't matter.

File: qiskit/extensions/hamiltonian_gate.py [hunk 8]
```diff
@@ -116,6 +117,17 @@ def _define(self):
...
+    @deprecate_func(
+        additional_msg=(
+            "All methods named `qasm` for classes derived from `Instruction` are deprecated."
+            " Stating that one in particular is not supported is redundant."
+        ),
+        since="0.25.0",
+    )
     def qasm(self):
         """Raise an error, as QASM is not defined for the HamiltonianGate."""
         raise ExtensionError("HamiltonianGate has no QASM definition.")
```
Comment (jakelishman) [inline qiskit/extensions/hamiltonian_gate.py:133]: This message doesn't seem to be worded correctly for the people who will see it.  It only applies to this particular subclass, and the people who see the message are users of the class, not definers, so they haven't done any subclassing, and so there's nothing actionable they can do.

It'll be fine to just give it the exact same error message as `Instruction.qasm`, and really, it's probably fine to just remove the override entirely, as mentioned above.

**Review by jakelishman — COMMENTED** [review]
I'll be glad to see the backs of these methods - one more place ticked off the list of instructions examining their `params` state.  Thanks for looking at it.

File: qiskit/circuit/library/standard_gates/x.py (hunk 1 above)
Comment (jlapeyre) [reply qiskit/circuit/library/standard_gates/x.py:595]: Huh? I think I only put that in because I was forced to (by the test suite, I thin). I recheck it. And revert if possible.

File: qiskit/circuit/quantumcircuit.py (hunk 2 above)
Comment (jlapeyre) [reply qiskit/circuit/quantumcircuit.py:1746]: I'll try to find another way to do this.

File: qiskit/circuit/quantumcircuit.py (hunk 3 above)
Comment (jlapeyre) [reply qiskit/circuit/quantumcircuit.py:2831]: I can try. I don't recall making that change myself. Maybe ruff?

File: qiskit/extensions/hamiltonian_gate.py (hunk 6 above)
Comment (jlapeyre) [reply qiskit/extensions/hamiltonian_gate.py:123]: I don't understand why it's magic. But if we don't need it, so much the better.

File: test/python/circuit/test_circuit_qasm.py (hunk 7 above)
Comment (jlapeyre) [reply test/python/circuit/test_circuit_qasm.py:273]: What specifically is not needed here? There are more than 800 tests for `DeprecationWarning` in the test suite. Is it something else? I'm not opposed to removing this.

File: qiskit/circuit/quantumcircuit.py (hunk 5 above)
Comment (jlapeyre) [reply qiskit/circuit/quantumcircuit.py:5232]: Maybe `qasm2_call` ?

File: qiskit/circuit/library/standard_gates/x.py (hunk 1 above)
Comment (jlapeyre) [reply qiskit/circuit/library/standard_gates/x.py]: Done in a790c38

File: qiskit/circuit/quantumcircuit.py (hunk 3 above)
Comment (jlapeyre) [reply qiskit/circuit/quantumcircuit.py]: Done in c062ca6

File: qiskit/circuit/quantumcircuit.py (hunk 2 above)
Comment (jlapeyre) [reply qiskit/circuit/quantumcircuit.py]: Removed in 68c1832

File: qiskit/extensions/hamiltonian_gate.py (hunk 6 above)
Comment (jlapeyre) [reply qiskit/extensions/hamiltonian_gate.py]: Removed in 68c1832.

File: qiskit/extensions/hamiltonian_gate.py (hunk 8 above)
Comment (jlapeyre) [reply qiskit/extensions/hamiltonian_gate.py]: Removed in 68c1832

File: qiskit/circuit/quantumcircuit.py [hunk 9]
```diff
@@ -5085,7 +5088,7 @@ def _qasm2_define_custom_operation(operation, existing_gate_names, gates_to_defi
                 instruction.operation, existing_gate_names, gates_to_define
             )
             bits_qasm = ",".join(qubit_labels[q] for q in instruction.qubits)
-            statements.append(f"{new_operation.qasm()} {bits_qasm};")
+            statements.append(f"{_instruction_qasm2(new_operation)} {bits_qasm};")
         body_qasm = " ".join(statements)
```
Comment (jlapeyre) [inline qiskit/circuit/quantumcircuit.py:5092]: Regarding "The ordering of these calls doesn't look right...". I made a couple of commits to simplify the code that show that all I've done is exchange a method call for a plain function call. However, I think this makes it clear that the call to `_instruction_qasm2` could be moved inside ` _qasm2_define_custom_operation`.  This would the code a bit less complex and add a bit more encapsulation. I think this may be what you are suggesting.

File: qiskit/circuit/quantumcircuit.py (hunk 4 above)
Comment (jlapeyre) [reply qiskit/circuit/quantumcircuit.py]: See https://github.com/Qiskit/qiskit-terra/pull/10399#pullrequestreview-1531079795 .

File: qiskit/circuit/quantumcircuit.py (hunk 5 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: `qasm2_call` sounds like the best option - agree that it's not perfect, but at the very least it's not misleading like `name_param`.

File: test/python/circuit/test_circuit_qasm.py (hunk 7 above)
Comment (jakelishman) [reply test/python/circuit/test_circuit_qasm.py:273]: I'm sorry - I don't know what I was thinking about.  The test is definitely fine, since this is an overridden method.

**Review by jakelishman — COMMENTED** [review]
Thanks for the changes, John.  I think once this is brought up to date with the current state of `main`, this will be ready to go, give or take a variable renaming which isn't even very important.

File: qiskit/circuit/quantumcircuit.py (hunk 9 above)
Comment (jlapeyre) [reply qiskit/circuit/quantumcircuit.py:5092]: Moving the hack down to a lower level feels worse in a sense. But, it is simpler and temporarily mutating something that should not change is distasteful. I didn't realize that it is more correct in the sense of your example. It's useful to make this clear

File: qiskit/circuit/quantumcircuit.py [hunk 10]
```diff
@@ -1646,6 +1648,7 @@ def qasm(
             MissingOptionalLibraryError: If pygments is not installed and ``formatted`` is
                 ``True``.
             QasmError: If circuit has free parameters.
+            CircuitError: If an operation that has no OpenQASM 2 representation is encountered.
```
[the remark this answers is not in the document]
Comment (jlapeyre) [reply qiskit/circuit/quantumcircuit.py:1651]: Must be cruft from multiple changes. I'll check it.

**Comment by jlapeyre** [comment]
I think it's worth making the small changes you mentioned @jakelishman.

What do you think of inlining `_instruction_qasm2` into `_qasm2_define_custom_operation`? In general I like breaking functions into smaller parts. But I'll have to check the logic to verify it's a good place to break it.

At a minimum `_instruction_qasm2`, if it is to remain a helper function, should be relocated next to `_qasm2_define_custom_operation` in the source.

File: qiskit/circuit/library/standard_gates/x.py [hunk 11]
```diff
@@ -568,6 +569,14 @@ def _define(self):
...
 
+    @deprecate_func(
+        additional_msg=(
+            "Correct exporting to OpenQASM 2 is the responsibility of a larger exporter; it cannot "
+            "safely be done on an object-by-object basis without context. No replacement will be "
+            "provided, because the premise is wrong."
+        ),
+        since="0.25.0",
+    )
     def qasm(self):
```
[the remark this answers is not in the document]
Comment (jlapeyre) [reply qiskit/circuit/library/standard_gates/x.py]: Done in 8961c66

File: qiskit/circuit/quantumcircuit.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jlapeyre) [reply qiskit/circuit/quantumcircuit.py]: Done in f78313c

File: qiskit/circuit/quantumcircuit.py [hunk 12]
```diff
@@ -44,6 +44,7 @@
 from qiskit.exceptions import QiskitError
 from qiskit.utils.multiprocessing import is_main_process
 from qiskit.circuit.instruction import Instruction
+import qiskit.circuit.instruction
```
Comment (jakelishman) [inline qiskit/circuit/quantumcircuit.py:47]: Is this needed?

**Review by jakelishman — COMMENTED** [review]
One rogue import aside, this looks good to merge.  Thanks for sticking with it and getting the whole thing written on short notice!

File: qiskit/circuit/quantumcircuit.py (hunk 12 above)
Comment (jlapeyre) [reply qiskit/circuit/quantumcircuit.py]: Removed in 9a2db41

**Review by jakelishman — APPROVED** [review]
